Use libmount more

keszybz opened the pull request:
This might be a solution for https://github.com/systemd/systemd/issues/12018, once https://github.com/karelzak/util-linux/issues/780 is resolved.

evverx commented:
This pull request **introduces 2 alerts** when merging 90b4063814be208942e355226f37536428d98330 into 949082ac214409365c01781fbf18ba08593918ff - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-d5c438894df59dc9f74160a649d93d3435f7cd9d)

**new alerts:**

* 1 for Missing header guard
* 1 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

keszybz commented:
Force-pushed to fix the two issues pointed out by LGTM.

poettering commented on src/shared/mount-util.c:
```diff
@@ -37,45 +35,37 @@ int umount_recursive(const char *prefix, int flags) {
-                        k = cunescape(path, UNESCAPE_RELAX, &p);
-                        if (k < 0)
-                                return k;
+                        if (cunescape(path, UNESCAPE_RELAX, &p) < 0)
```
uh? my assumption was that libmount would do the unescaping for us, so this can go, no?

poettering commented on src/shared/mount-util.c:
```diff
@@ -37,45 +35,37 @@ int umount_recursive(const char *prefix, int flags) {
+                table = mnt_new_table();
+                iter = mnt_new_iter(MNT_ITER_FORWARD);
+                if (!table || !iter)
+                        return log_oom();
```
this function currently doesn't log, it's "library-like". I don't think we should change that

poettering commented on src/shared/mount-util.c:
```diff
@@ -37,45 +35,37 @@ int umount_recursive(const char *prefix, int flags) {
-                (void) __fsetlocking(proc_self_mountinfo, FSETLOCKING_BYCALLER);
+                r = mnt_table_parse_mtab(table, NULL);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse /proc/self/mountinfo: %m");
```
as above, please downgrade to LOG_DEBUG

poettering commented on src/shared/mount-util.c:
```diff
@@ -37,45 +35,37 @@ int umount_recursive(const char *prefix, int flags) {
+                        if (r == 1)
+                                break;
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to get next entry from /proc/self/mountinfo: %m");
```
downgrade to debug

poettering commented on src/shared/mount-util.c:
```diff
@@ -153,37 +145,38 @@ int bind_remount_recursive_with_mountinfo(
+                table = mnt_new_table();
+                iter = mnt_new_iter(MNT_ITER_FORWARD);
+                if (!table || !iter)
+                        return log_oom();
```
logging

poettering commented on src/shared/mount-util.c:
```diff
@@ -153,37 +145,38 @@ int bind_remount_recursive_with_mountinfo(
 
+                r = mnt_table_parse_stream(table, proc_self_mountinfo, "/proc/self/mountinfo");
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse /proc/self/mountinfo: %m");
```
logging

poettering commented on src/shared/mount-util.c:
```diff
@@ -153,37 +145,38 @@ int bind_remount_recursive_with_mountinfo(
+                        if (r == 1)
+                                break;
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to get next entry from /proc/self/mountinfo: %m");
```
logging

poettering commented on src/shared/mount-util.c:
```diff
@@ -153,37 +145,38 @@ int bind_remount_recursive_with_mountinfo(
+                                continue;
+
+                        if (cunescape(path, UNESCAPE_RELAX, &p) < 0)
+                                return log_oom();
```
logging and do we really need to escape here?

keszybz commented:
You're right about the unescaping. I remember this was discussed when the current code was being merged, but for some reason the unescaping was left in, even though it shouldn't. Did maybe libmount behave differently in the past? Anyway, I added a patch to remove the unescaping in all places it was done, and adjusted the new patches to not do it. PTAL.

poettering commented on src/core/mount.c:
```diff
@@ -11,7 +11,6 @@
 #include "dbus-mount.h"
 #include "dbus-unit.h"
```
this commit should probably be added to v242 still?

keszybz replied:
Maybe the first three (refactoring, test, and cunescape removal)? I can submit separate PR.

poettering commented on src/shared/mount-util.c:
```diff
@@ -37,55 +34,44 @@ int umount_recursive(const char *prefix, int flags) {
+                iter = mnt_new_iter(MNT_ITER_FORWARD);
+                if (!table || !iter)
+                        return log_oom();
 
```
still logs at non-debug level

keszybz replied:
This one is different, it was logging at non-debug level before, I just kept the behaviour.

poettering replied:
but please. let's fix this when touching that

keszybz replied:
It seems you were right, and I added logging here, but then removed it in one of the later commits. I now fixed this during the rebase to have `return -ENOMEM`.

poettering commented on src/shared/mount-util.c:
```diff
@@ -219,15 +202,12 @@ int bind_remount_recursive_with_mountinfo(
-                                        continue;
+                        if (!set_contains(done, path)) {
+                                r = set_put_strdup(todo, path);
                                 if (r < 0)
```
needs to handle EEXIST gracefully, as before

keszybz replied:
`set_put_strdup` return 0 if the entry exists.

poettering commented on src/mount/mount-tool.c:
```diff
@@ -713,65 +714,59 @@ static int start_transient_automount(
-        for (;;) {
-                _cleanup_free_ char *path = NULL, *where = NULL, *dev = NULL;
-                int r;
+        r = mnt_table_parse_mtab(table, NULL);
```
maybe we should add a helper for this to libmount-util:

```c
int mnt_table_proc_self_mount_info(struct libmnt_table **ret);
```

which allocates the table and opens it in one go. and maybe even go one step further and also allocate an iter at the same time and return that. After all we currently always do these three steps separately, always the same way

keszybz commented:
Force pushed with a bugfix (wrong return value on error), and the suggested change to add a helper function. I also added one more commit to relax the test case which fails in CI.

keszybz commented:
I think the CI failure is again caused by old libmount version. I hope it'll pass now.

poettering commented on src/shared/libmount-util.h:
```diff
@@ -1,10 +1,43 @@
 DEFINE_TRIVIAL_CLEANUP_FUNC(struct libmnt_table*, mnt_free_table);
 DEFINE_TRIVIAL_CLEANUP_FUNC(struct libmnt_iter*, mnt_free_iter);
+
+static inline int libmount_parse(
```
hmm, static inline? I think this should just be a regular function, it's complicated enough to warrant that.

keszybz replied:
This was explained in the commit message that adds the function. I'll repush for the other things, and then I'll try to make the explanation clearer.

poettering commented on src/shared/libmount-util.h:
```diff
@@ -1,10 +1,43 @@
+
+        if (source)
+                /* Older libmount seems to require this. */
+                assert(path);
```
`assert(!source || path)`?

poettering commented on src/test/test-libmount.c:
```diff
@@ -28,7 +30,12 @@ static void test_libmount_unescaping_one(
 
-        assert_se(mnt_table_next_fs(table, iter, &fs) == 0);
+        r = mnt_table_next_fs(table, iter, &fs);
+        if (r < 0 && may_fail) {
```
can you add a comment here maybe, that says "we allow this to fail in some tests, see below", or so?

keszybz commented:
Updated:
- the test is updated to reflect the changes in libmount that were merged upstream in the meantime
- things are skipped more consistently in tests, and the requested comment has been added
- the log_oom() thing pointed out in review is fixed
- assert() is updated as requested

mrc0mmand commented on src/shared/mount-util.c:
```diff
@@ -37,55 +34,39 @@ int umount_recursive(const char *prefix, int flags) {
-                                        break;
+                        struct libmnt_fs *fs;
+                        const char *path;
+                        _cleanup_free_ char *p = NULL;
```
To make Travis (clang) happy, this unused variable should be dropped.

mrc0mmand commented on src/shared/mount-util.c:
```diff
@@ -155,56 +138,46 @@ int bind_remount_recursive_with_mountinfo(
+                for (;;) {
+                        struct libmnt_fs *fs;
+                        const char *path, *type;
+                        _cleanup_free_ char *p = NULL;
```
Ditto.

keszybz commented:
bionic-i386 fails in two places: test/TEST-29-UDEV-ID_RENAMING test/TEST-30-ONCLOCKCHANGE. 
TEST-29 fails with "systemd-testsuite login: Failed to wait for daemon to reply: Connection timed out", might be just a timeout.
TEST-30 is known flakey.
I'll force push to fix the errors pointed out by @mrc0mmand.

mrc0mmand commented:
The Travis CI integration is having an aneurysm, so I triggered it manually: https://travis-ci.org/systemd/systemd/builds/517648222

poettering commented on src/shared/mount-util.c:
```diff
@@ -140,14 +135,9 @@ int bind_remount_recursive_with_mountinfo(
                 rewind(proc_self_mountinfo);
 
-                r = mnt_table_parse_stream(table, proc_self_mountinfo, "/proc/self/mountinfo");
+                r = libmount_parse("/proc/self/mountinfo", proc_self_mountinfo, &table, &iter);
```
hmm, so i wonder about the two uses in bind_remount_recursive_with_mountinfo() and umount_recursive(). We use this when setting up a namespace environment for a service. i.e. we operate on a different root than the real one. libmount is unlikely to understand that though and will merge data from  /run/mount/utab into what is read here, but that's data about host mounts, not about mounts in the little namespaced environment we are setting up. Hence, we should somehow massage libmount into not loading utab meta info here if we want to use it in an entirely generic function like these two that should make no assumptions on the context they are called in.

@karelzak can you comment on this maybe? is there a way we can use the libmount mnt_table_xyz API without it reading utab for such namespaced environments?

karelzak replied:
> @karelzak can you comment on this maybe? is there a way we can use the libmount mnt_table_xyz API without it reading utab for such namespaced environments?

Only mnt_table_parse_mtab() reads utab userspace junk.

So, all you need is to use mnt_table_parse_file(), mnt_new_table_from_file() or mnt_table_parse_stream() with "/proc/self/mountinfo" as path to the table.

And for libmount_monitor API -- the utab is enabled only if explicitly requested by  mnt_monitor_enable_userspace().
